mmv1/third_party/terraform/resources/resource_pubsub_schema.go.erb

```diff
@@ -0,0 +1,629 @@
+	}
+
+	if err := d.Set("project", project); err != nil {
+		return fmt.Errorf("Error reading Schema: %s", err)
+	}
+
+	if err != nil {
+		return err
+	}
+	ss := res["schemas"].([]interface{})
+	var tss []interface{}
+	df := false
+	// listRevisions returns schemas in reverse chronological order.
+	for i := len(ss) - 1; i >= 0; i-- {
+		rs := ss[i].(map[string]interface{})
+		if d.Get("use_definition").(bool) && !df {
+			// The oldest revision should be the schema-wide "definition" if
+			// "definition" is used.
+			d.Set("definition", rs["definition"])
+			df = true
+		} else {
+			ts := make(map[string]interface{})
+			ts["definition"] = rs["definition"]
+			ts["revision_id"] = rs["revisionId"]
+			ts["revision_create_time"] = rs["revisionCreateTime"]
+			tss = append(tss, ts)
+		}
+		d.Set("name", NameFromSelfLinkStateFunc(strings.Split(rs["name"].(string), "@")[0]))
+		d.Set("type", rs["type"])
+	}
+	d.Set("revision", tss)
+	return nil
+}
+
+func resourcePubsubSchemaUpdate(d *schema.ResourceData, meta interface{}) error {
+	config := meta.(*transport_tpg.Config)
+	userAgent, err := tpgresource.GenerateUserAgentString(d, config.UserAgent)
+	if err != nil {
+		return err
+	}
+
+	billingProject := ""
+
+	project, err := tpgresource.GetProject(d, config)
+	if err != nil {
+		return fmt.Errorf("Error fetching project for Schema: %s", err)
+	}
+	billingProject = project
+
+	rObj := make(map[string]interface{})
+	rObj["name"] = d.Id()
+	obj := make(map[string]interface{})
+	typeProp := d.Get("type")
+	if err != nil {
+		return err
+	} else if v, ok := d.GetOkExists("type"); !isEmptyValue(reflect.ValueOf(v)) && (ok || !reflect.DeepEqual(v, typeProp)) {
+		obj["type"] = typeProp
+	}
+	obj["name"] = d.Id()
+
+	old, new := d.GetChange("revision")
+	lold := old.([]interface{})
+	lnew := new.([]interface{})
+
+	adds, dels := computeRevisionDiffs(lold, lnew)
+
+	oldr := make(map[string]bool)
+	for _, r := range lold {
+		rm := r.(map[string]interface{})
+		oldr[rm["revision_id"].(string)] = true
+	}
+
+	revC := len(lold)
+	if d.Get("definition") != "" {
+		revC++
+	}
+
+	d.Partial(true)
+	dObj := make(map[string]interface{})
+	dObj["name"] = d.Id()
+
+	dCount := 0
+	aCount := 0
+
+	// Alternate deletes and commmits of revisions in order to keep the number of
+	// revisions within the allowed bounds. We initially do these in the order
+	// delete/add, but if we are going to delete the last revision, we switch to
+	// add/delete.
+	for dCount < len(dels) || aCount < len(adds) {
+		if dCount < len(dels) && revC > 1 {
+			deletePubsubSchemaRevision(dels[dCount], dObj, d, config, userAgent, billingProject)
+			dCount++
+			revC--
+		}
+		if aCount < len(adds) {
+			obj["definition"] = adds[aCount]
+			rObj["schema"] = obj
+			err = commitPubsubSchemaRevision(rObj, d, config, userAgent, billingProject)
+			if err != nil {
+				return err
+			}
+			aCount++
+			revC++
+		}
+	}
```

Comment:
Is there a different approach that could be taken for this code? Perhaps deleting everything first and then adding them all later? I'm not sure what API restrictions there are besides the limit of 20 revisions- e.g. is there a minimum number of revisions?

Reply:
There is also a limitation that there has to be at least one revision. So we can't delete all first and then add. And because of the max 20 revisions, we can't add and then remove. I tried a few different implementations and this one ended up seeming to be the cleanest.